app/src/main/java/com/portfolio/tasky/entry/fragments/LoginFragment.kt

```diff
@@ -141,18 +140,30 @@ class LoginFragment : Fragment(), FragmentInflater by FragmentInflaterImpl(), Te
             )
         )
     }
-    /**
-     * Necessary for navigating to RegistrationFragment.
-     * Currently not called
-     */
+
+    private fun startLogin() {
+        val email = viewBinding.etEmail.subLayout.etInput.text.toString()
+        val password = viewBinding.etPassword.subLayout.etInput.text.toString()
+
+        lifecycleScope.launch(Dispatchers.IO + coroutineExceptionHandler){
+            viewModel.login(AuthenticationRequest(email, password))
+        }
+
+        viewModel.loginObserver.observe(viewLifecycleOwner){
+            if(it != null){
+                it.token?.let { it1 -> LoginUtils.setAccessToken(it1)}
+                Toast.makeText(activity, "Token ${it.token}", Toast.LENGTH_SHORT).show()
+            }
+        }
```

Comment:
I'd avoid putting such data related stuff (like saving something in persistent storage) in the UI. Keep it as dumb as possible and only display something based on what you get from the ViewModel.

The following allows you to easily access user data from any layer in your app while being able to also easily swap out preferences implementations. The only thing needed for that would be to swap out the shared preferences object in the `UserPreferencesImpl` constructor with let's say a DataStore instance.

*Domain model*
```
data class AuthenticatedUser(
    val fullName: String,
    val email: String,
    val token: String
)
```

*Domain abstraction*
```
interface UserPreferences {
    fun saveAuthenticatedUser(user: AuthenticatedUser)
    fun getAuthenticatedUser(): AuthenticatedUser?
    fun clearPreferences()
}
```

*Data implementation*
```
class UserPreferencesImpl(
    private val sharedPref: SharedPreferences
): UserPreferences {
    override fun saveAuthenticatedUser(user: AuthenticatedUser) {
        // Save JSON of user in shared pref
    }

    override fun getAuthenticatedUser(): AuthenticatedUser? { ... }

    override fun clearPreferences() { ... }
}
```

In your Dagger module:
```
@Provides
@Singleton
fun provideUserPreferences(sharedPreferences: SharedPreferences): UserPreferences {
    // For sensitive user data like tokens, I'd use EncryptedSharedPreferences
    return UserPreferencesImpl(sharedPreferences)
}
```

Then, for example in any ViewModel:
```
@HiltViewModel
class MyViewModel @Inject constructor(
    private val userPreferences: UserPreferences
): ViewModel() {
    // Use userPreferences
}
```

Since your ViewModel knows if the login was successful or not, it can then use `userPreferences` to save the user data 👍 This way, you keep your preferences implementation data specific while being able to use it across the app and easily swap out how preferences are implemented.

Reply:
Great info!

Done